File: prefix_sum/floored_pairs_sum/floored_pairs_sum.cpp
#include <vector>
#include <unordered_set>
using namespace std;

int sumFlooredPairs(vector<int> &nums)
{ // LeetCode Q.1862.
    int maxNum = *max_element(nums.begin(), nums.end());
    vector<int> numsCounts(maxNum + 1, 0); // Idx i stores count of num i.

    unordered_set<int> distinctNums;
    for (auto num : nums)
    {
        numsCounts[num] += 1;
        distinctNums.insert(num);
    }

    vector<int> prefixCounts;
    for (int idx = 0; idx < numsCounts.size(); idx++)
    {
        if (idx == 0)
            prefixCounts.push_back(numsCounts[idx]);

        else
            prefixCounts.push_back(prefixCounts.back() + numsCounts[idx]);
    }

    long long flooredPairsSum = 0, modulo = pow(10, 9) + 7;
    for (auto num : distinctNums)
    {
        int maxMultiple = maxNum / num;

        for (int multiple = 1; multiple <= maxMultiple; multiple++)
        {
            long long count = 0, product = num * multiple;

            if (multiple == maxMultiple)
                count += prefixCounts.back();

            else
                count += prefixCounts[product + num - 1];

            count -= prefixCounts[product - 1];

            long long increment = numsCounts[num] % modulo;
            increment *= multiple % modulo;
            increment *= count % modulo;

            flooredPairsSum += increment;
            flooredPairsSum %= modulo;
        }
    }

    return flooredPairsSum;
}

File: prefix_sum/1862_hard/floored_pairs_sum.py

def sum_floored_pairs(nums: list[int]) -> int:  # LeetCode Q.1862.
    max_num = max(nums)
    nums2counts = [0] * (max_num + 1)  # Idx i stores count of num i. Min num >= 0.
    for num in nums:
        nums2counts[num] += 1

    prefix_counts = []
    for count in nums2counts:
        if not prefix_counts:
            prefix_counts.append(count)
            continue
        prefix_counts.append(prefix_counts[-1] + count)

    floored_pairs_sum, modulo = 0, 10 ** 9 + 7
    distinct_nums = set(nums)
    for num in distinct_nums:
        max_multiple = max_num // num
        for multiple in range(1, max_multiple + 1):
            product = num * multiple

            if multiple == max_multiple:
                count = prefix_counts[-1]

            else:
                count = prefix_counts[product + num - 1]

            count -= prefix_counts[product - 1]

            floored_pairs_sum += nums2counts[num] * multiple * count
            floored_pairs_sum %= modulo

    return floored_pairs_sum
